### DISCUSSION TITLE
cannot render cone and head together

### DISCUSSION THREAD

**Discussion by guozhi72** [Kitware/trame#271] [Q&A]
I set up two pipelines one is to render cone, another is to render contour of head. But unless I use localview and remote view seperately, I will see same volume. 

So what is wrong with my understanding? 

thanks

```python
from trame.widgets import vuetify, html

from trame.widgets import vuetify, vtk as vtk_widgets
from trame.ui.vuetify import SinglePageLayout
from vtkmodules.vtkFiltersSources import vtkConeSource
from vtkmodules.vtkIOXML import vtkXMLImageDataReader
from vtkmodules.vtkFiltersCore import vtkContourFilter
from vtkmodules.vtkRenderingCore import (
    vtkRenderer,
    vtkRenderWindow,
    vtkRenderWindowInteractor,
    vtkPolyDataMapper,
    vtkActor,
)

# VTK factory initialization
from vtkmodules.vtkInteractionStyle import vtkInteractorStyleSwitch  # noqa
import vtkmodules.vtkRenderingOpenGL2  # noqa



class MyVolume:

    def __init__(self, server):
        self.server = server
        self.state = server.state
        self.ctrl = server.controller
        state = self.state
        self.message = None

        ##### cone

        self.renderer_2 = vtkRenderer()
        self.renderWindow_2 = vtkRenderWindow()
        self.renderWindow_2.AddRenderer(self.renderer_2)

        self.renderWindowInteractor_2 = vtkRenderWindowInteractor()
        self.renderWindowInteractor_2.SetRenderWindow(self.renderWindow_2)
        self.renderWindowInteractor_2.GetInteractorStyle().SetCurrentStyleToTrackballCamera()

        self.cone_source_2 = vtkConeSource()
        self.mapper_2 = vtkPolyDataMapper()
        self.mapper_2.SetInputConnection(self.cone_source_2.GetOutputPort())
        self.actor_2 = vtkActor()
        self.actor_2.SetMapper(self.mapper_2)

        self.renderer_2.AddActor(self.actor_2)
        self.renderer_2.ResetCamera()


        # -----------------------------------------------------------------------------
        # head contour
        # -----------------------------------------------------------------------------


        head_vti = r"C:\dev_azure\rendering_trame\Data\head.vti"
        self.reader = vtkXMLImageDataReader()
        self.reader.SetFileName(head_vti)
        self.reader.Update()

        self.contour = vtkContourFilter()
        self.contour.SetInputConnection(self.reader.GetOutputPort())
        self.contour.SetComputeNormals(1)
        self.contour.SetComputeScalars(0)

        # Extract data range => Update store/state
        data_range = self.reader.GetOutput().GetPointData().GetScalars().GetRange()
        contour_value = 0.5 * (data_range[0] + data_range[1])
        state.data_range = data_range

        # Configure contour with valid values
        self.contour.SetNumberOfContours(1)
        self.contour.SetValue(0, contour_value)

        # Rendering setup
        self.renderer = vtkRenderer()
        self.renderWindow = vtkRenderWindow()
        self.renderWindow.AddRenderer(self.renderer)

        self.renderWindowInteractor = vtkRenderWindowInteractor()
        self.renderWindowInteractor.SetRenderWindow(self.renderWindow)
        self.renderWindowInteractor.GetInteractorStyle().SetCurrentStyleToTrackballCamera()

        self.mapper = vtkPolyDataMapper()
        self.actor = vtkActor()
        self.mapper.SetInputConnection(self.contour.GetOutputPort())
        self.actor.SetMapper(self.mapper)
        self.renderer.AddActor(self.actor)
        self.renderer.ResetCamera()
        # renderWindow.Render()





    def update_tab(self, tab, **kwargs):
        print(f"message===={tab}")
        self.message = tab


    def layout(self):
        # with RouterViewLayout(self.server, "/"):
        with vuetify.VCard():
            vuetify.VCardTitle("This is volume page")
            vuetify.VBtn(
                "volume view button",
                click=self.send_to_server1
            )
        with vuetify.VContainer(key='345'):
            view = vtk_widgets.VtkLocalView(self.renderWindow, style="min-height: 600px; max-width: 600px")
                # vuetify.VCardTitle("This is home")
            self.ctrl.view_update = view.update
            self.ctrl.view_reset_camera = view.reset_camera
            self.ctrl.on_server_ready.add(view.update)
        with vuetify.VContainer(key='123'):

            view = vtk_widgets.VtkLocalView(self.renderWindow_2, style="min-height: 600px; max-width: 600px")

    def send_to_server1(self):

        print(f"11111 volume button pressed")


    def on_button_123(self, message):
        print('123 is pressed')


```

**Comment by jourdain** [comment]
Adding the ref= should solve the issue

```python
# [...]
       with vuetify.VContainer(key='345'):
            view = vtk_widgets.VtkLocalView(self.renderWindow, ref="view_cone", style="min-height: 600px; max-width: 600px")
                # vuetify.VCardTitle("This is home")
            self.ctrl.view_update = view.update
            self.ctrl.view_reset_camera = view.reset_camera
            self.ctrl.on_server_ready.add(view.update)
        with vuetify.VContainer(key='123'):

            view = vtk_widgets.VtkLocalView(self.renderWindow_2, ref="view_volume", style="min-height: 600px; max-width: 600px")
```

**Reply by guozhi72** [reply]
Jourdian:

Thanks you so much, that help. Can you teach me if there is any way I can figure out this addig ref stuff by my self.
having a lot of trouble to server side rending using vtk/paraview, and this new trame. Just have no clue, don't know how to solve the problem.

Not like anuglar, Vue, can find from anywhere on the internet

**Comment by jourdain** [comment]
`ref=` are vue.js things to ID an element so you can call methods on them. Those advanced trame components do need to be able to call methods on the JS side. So if you have more than 1, you need to make sure you provide a unique name for each of them.

By default, that ref arg is provided and set to "view" which does not work once you have more than 1... It is part of the args of the constructor which you can see in [the API doc](https://trame.readthedocs.io/en/latest/index.html).

**Reply by guozhi72** [reply]
Hi, Jourdain:

Thanks, but do struggling to understand the documentation, For example,  vtk_widgets.VtkRemoteView(self.renderWindow,
                                             ref="view_volume_1", style="min-height: 600px; max-width: 600px")
vtkRemoteView is python class, but style is html stuff.
1, can I apply all html stuff form vue to here?
2, if hue, v-if is for conditional rendering, how can I do conditional rendering for in trame

**Reply by jourdain** [reply]
The python here is to build the vue template so technically it is all HTML stuff there. 
Maybe you should start with the [tutorial](https://kitware.github.io/trame/docs/tutorial.html) and [course](https://kitware.github.io/trame/docs/course_intro.html).

**Reply by guozhi72** [reply]
Hi, jourdain:

I went through trame tutorial many times. But still feel a lot of difiicult.
1, I developed some angular code for slice and volume rendering using angular and vtk.js. Intially have some difficulty to connect angular with vtk.js. But after setup connection, first UI development is so flexible, stuck anywhere, gooogled and get solution.
2, but using trame, feel really difficult,
3, for example, conditional rendering using angular or vue, I can find how to do it immediately, but using trame, struggle a hours, don't know how to start it all all. where and how I can use v-if??

**Reply by jourdain** [reply]
You should go with a trame server and angular client if you feel it is simpler. Trame is not forcing you to do anything. 

And your v-if/v-show can be used everywhere... Trame is just exposing everything that vue.js has to offer. Nothing more or less.
Again if you understand vue.js, with trame is exactly the same thing... Just replace invalid char (-:.) by _ and you should be good...

And if you don't want to describe your vue template using Python, you can do it in plain string too... 

```python
vuetify.VBtn("Toggle visibility", click=toggle_visibility)
with vuetify.VCard(v_show=("show_card", False)):
    vuetify.VCardTitle("Hello")


def toggle_visibility():
     server.state.show_card = not server.state.show_card
```

**Reply by guozhi72** [reply]
Hi, Jourdain:

Thanks for demo how to trigger show on/off using trame, but still have several questions:
1, how do I know I should use v_show. If use vue, I search, I know if should use v-if in a second. If use angular, I know I should use ngIf very quickly. But using trame, how can I find out I should use v_show? Can you teach me the process to find out v_show?

2, I tried this toggle using below code. If I start with v_show=True, I can turn if off/one, works perfectly. But if I start with False, initially the there is no space to render volume view, but after v_show is True, I saw new space created, but only white

        with vuetify.VContainer(v_show=("show_volume", True)):
            view = vtk_widgets.VtkRemoteView(self.renderWindow,
                                             ref="view_volume_1", style="min-height: 600px; max-width: 600px")
                # vuetify.VCardTitle("This is home")

####################################### coded
from trame.widgets import vuetify, html

from trame.widgets import vuetify, vtk as vtk_widgets
from trame.ui.vuetify import SinglePageLayout
from vtkmodules.vtkFiltersSources import vtkConeSource
from vtkmodules.vtkIOXML import vtkXMLImageDataReader
from vtkmodules.vtkFiltersCore import vtkContourFilter
from vtkmodules.vtkRenderingCore import (
    vtkRenderer,
    vtkRenderWindow,
    vtkRenderWindowInteractor,
    vtkPolyDataMapper,
    vtkActor,
)

# VTK factory initialization
from vtkmodules.vtkInteractionStyle import vtkInteractorStyleSwitch  # noqa
import vtkmodules.vtkRenderingOpenGL2  # noqa



class MyVolume:

    def __init__(self, server):
        self.display_head = False
        self.server = server
        self.state = server.state
        self.ctrl = server.controller
        state = self.state
        self.message = None

        ##### cone

        self.renderer_2 = vtkRenderer()
        self.renderWindow_2 = vtkRenderWindow()
        self.renderWindow_2.AddRenderer(self.renderer_2)

        self.renderWindowInteractor_2 = vtkRenderWindowInteractor()
        self.renderWindowInteractor_2.SetRenderWindow(self.renderWindow_2)
        self.renderWindowInteractor_2.GetInteractorStyle().SetCurrentStyleToTrackballCamera()

        self.cone_source_2 = vtkConeSource()
        self.mapper_2 = vtkPolyDataMapper()
        self.mapper_2.SetInputConnection(self.cone_source_2.GetOutputPort())
        self.actor_2 = vtkActor()
        self.actor_2.SetMapper(self.mapper_2)

        self.renderer_2.AddActor(self.actor_2)
        self.renderer_2.ResetCamera()


        # -----------------------------------------------------------------------------
        # VTK pipeline
        # -----------------------------------------------------------------------------


        file_path = r"C:\dev_azure\rendering_trame\Data\foot.vti"
        reader = vtkXMLImageDataReader()
        reader.SetFileName(file_path)
        reader.Update()

        contour = vtkContourFilter()
        contour.SetInputConnection(reader.GetOutputPort())
        contour.SetComputeNormals(1)
        contour.SetComputeScalars(0)

        # Extract data range => Update store/state
        data_range = reader.GetOutput().GetPointData().GetScalars().GetRange()
        contour_value = 0.5 * (data_range[0] + data_range[1])
        state.data_range = data_range

        # Configure contour with valid values
        contour.SetNumberOfContours(1)
        contour.SetValue(0, contour_value)

        # Rendering setup
        renderer = vtkRenderer()
        renderWindow = vtkRenderWindow()
        renderWindow.AddRenderer(renderer)

        renderWindowInteractor = vtkRenderWindowInteractor()
        renderWindowInteractor.SetRenderWindow(renderWindow)
        renderWindowInteractor.GetInteractorStyle().SetCurrentStyleToTrackballCamera()

        mapper = vtkPolyDataMapper()
        actor = vtkActor()
        mapper.SetInputConnection(contour.GetOutputPort())
        actor.SetMapper(mapper)
        renderer.AddActor(actor)
        renderer.ResetCamera()
        # renderWindow.Render()

        self.renderWindow = renderWindow
        self.reader = reader







    def update_tab(self, tab, **kwargs):
        print(f"message===={tab}")
        self.message = tab


    def layout(self):
        # with RouterViewLayout(self.server, "/"):
        with vuetify.VCard():
            vuetify.VCardTitle("This is volume page")
            vuetify.VBtn(
                "volume view button",
                click=self.send_to_server1
            )
            vuetify.VBtn("Toggle visibility", click=self.toggle_visibility)
        with vuetify.VContainer(v_show=("show_volume", True)):
            view = vtk_widgets.VtkRemoteView(self.renderWindow,
                                             ref="view_volume_1", style="min-height: 600px; max-width: 600px")
                # vuetify.VCardTitle("This is home")
            self.ctrl.view_update = view.update
            self.ctrl.view_reset_camera = view.reset_camera
            self.ctrl.on_server_ready.add(view.update)
        with vuetify.VContainer(key='123'):

            view = vtk_widgets.VtkRemoteView(self.renderWindow_2, ref="cone_volume_2", style="min-height: 600px; max-width: 600px")

    def toggle_visibility(self):
        self.state.show_volume = not self.state.show_volume
        view = vtk_widgets.VtkRemoteView(self.renderWindow,
                                         ref="view_volume_1", style="min-height: 600px; max-width: 600px")
        self.ctrl.view_update = view.update
        self.ctrl.view_reset_camera = view.reset_camera
        self.ctrl.on_server_ready.add(view.update)

**Reply by guozhi72** [reply]
Hi, Jourdain:

What I feel angular is simple, I mean using vtk.js as client side rendering.
For vtk, paraview server side rendering, I spent more than a week, could not make a single sample worked using anular as client.

**Reply by jourdain** [reply]
https://vuejs.org/guide/essentials/template-syntax.html

**Reply by guozhi72** [reply]
the link showed v_if, but how can I find out I should use v_show? Basically, feel almost have no clue to figure out if I want to do anything in mind. But that may be due the less number of users. So difficicult to search. Angular or vue have thouthand times more users.
<p v-if="seen">Now you see me</p>

**Reply by jourdain** [reply]
So you are saying that making a client/server application is actually more difficult in Angular.
Then with trame you can't wrap your head around the vue.js and Python/trame relationship. 
But if you are dealing with a client/server code, you need to get a handle on that... There is no documentation on how to do that.

It seems that you are trying to find answers in trame API when actually the documentation belong to [vue.js templating](https://vuejs.org/guide/essentials/template-syntax.html). Trame is just providing a Python wrapper to that vue.js templating infrastructure while allowing to bind easily state (that lives in both client and server side) along with Python method call.  That's only what trame is doing. Nothing more but also nothing less.

As opposed to Streamlit, Dash or Solara, which only expose their own version of HTML elements, trame is exposing the full templating infrastructure of vue.js with no limitation what so ever. Any vue.js library can be used out of the box. But we've created some basic ones that can easily be used in Python to handle plotly, vtk, paraview, matplotlib and more. But in truth, they could have been left in plain JavaScript/Vue.js and still be used within trame. But by making them Python packages, they became fully integrated and allow less language jumping.

**Reply by guozhi72** [reply]
Thanks for your help. I am not really sure where I am blocked. but this link ( https://vuejs.org/guide/essentials/template-syntax.html) is for html template, I used angular, so quite familiar with this part.
I think, I am confused it, I couldnot find a way how to code those htm templates in python trame.

If use trame as server only, vue as client, it seems that have big flexibility. But I only see some example from you using paraview at server, vue as client. I did not see where using trame as server.

I saw you post, trame under neath use paraview. But how to use trame as server, I have no idea. 
Trame can easily get started, but difficult to customize the code from my experiment. Using vue client/ paraview as server, even setting up is very difficult.

**Reply by jourdain** [reply]
Trame give you 100% of the capability of doing everything in JS except you don't need to build the JS code yourself. So it is as powerful. You still didn't understand the 3 rules needed to cover the full vue.js capabilities. Because that template syntax is truly all you need to know along with the 3 rules on how it get written in Python rather than html.

For the JS client with trame server, it is like ParaViewWeb which means you have application as reference (no doc). Just need to manage your WebSocket and RPC yourself. A lower level but you are in control if you prefer that.

**Comment by guozhi72** [comment]
Jourdain:

Thanks a lot. With your help, in one day, I made a lot of progress. As a poc project, I am so glad. But just thinking, how can I figure out without your help.

Best

**Reply by jourdain** [reply]
Do a vue.js training? Look at the examples? Take the tutorial and course? I'm not sure, it really depend on what is blocking you.

**Comment by jourdain** [comment]
The conditional rendering is addressed here in vue.js: https://vuejs.org/guide/essentials/conditional.html

**Reply by guozhi72** [reply]
Yes, but that is use v-if. The example you provide use v_show( again, I still did not know how to find this v_show if you don't tell me). But now I have issue with v_show. If I start with v_show = False, I couldn't see the rendered volume. But if I start with v_show= True, then everything works fine. I post the code before

**Reply by jourdain** [reply]
That part is covered in the reference of the course (like many of your previous questions). You are very much getting confused between static template vs dynamic. If you say `html.Div(something=True, something_else=False)` this is pure static with no reactivity set. For dynamic, you need to define variable `html.Div(something=("something_value", True), something_else=("something_else", False)` then you can change those values to `state.something_value = False`.

**Reply by jourdain** [reply]
That doc cover both `v-if` and `v-show` and they all (meaning everything) works the same way...

**Reply by guozhi72** [reply]
I agreed, I am confused with many thing. And for trame not many user search it, so even more difficult at beginning

**Comment by guozhi72** [comment]
Can trame have more practical example? For example, 
1, trame has routing example, but only rendering VBtn or something like this, very simple thing.
2, when first open rending page, the page should not render anything, After User click rending or upload a file, start to rendering.
3, trame example, all volume rendering is hardcode.( when I try to start from empty view, at least to me, it is not straight forward)
4, also multiple tabs for rendering will be very helpful.
5, code seperation.

**Reply by jourdain** [reply]
1. Showing a button or something else should not be different.
2. I don't understand why that should be an example
3. That seems to be a related to VTK but not trame
4. https://v2.vuetifyjs.com/en/components/tabs/ 
5. The [cookiecutter](https://github.com/kitware/trame-cookiecutter) provide a taste of it. But as usual it is up to the user to do what pleases him.

**Comment by guozhi72** [comment]
Like below code:

1, If I use vtkLocalView, I can set intial value of v_show to be either True/False, toggle it, works perfectly.

2, But if I use vtkRemoteView, only works if I set intial vlaue to be True. Othewise, it won't work.

3, At this stage, I totoal have no idea on how to solve it. Any suggestions? 

```python
from trame.widgets import vuetify, vtk as vtk_widgets
from trame.ui.vuetify import SinglePageLayout
from vtkmodules.vtkFiltersSources import vtkConeSource
from vtkmodules.vtkIOXML import vtkXMLImageDataReader
from vtkmodules.vtkFiltersCore import vtkContourFilter
from vtkmodules.vtkRenderingCore import (
    vtkRenderer,
    vtkRenderWindow,
    vtkRenderWindowInteractor,
    vtkPolyDataMapper,
    vtkActor,
)

# VTK factory initialization
from vtkmodules.vtkInteractionStyle import vtkInteractorStyleSwitch  # noqa
import vtkmodules.vtkRenderingOpenGL2  # noqa



class MyVolume:

    def __init__(self, server):
        self.display_head = False
        self.server = server
        self.state = server.state
        self.ctrl = server.controller
        state = self.state
        self.message = None




        # -----------------------------------------------------------------------------
        # VTK pipeline
        # -----------------------------------------------------------------------------


        file_path = r"C:\dev_azure\rendering_trame\Data\foot.vti"
        reader = vtkXMLImageDataReader()
        reader.SetFileName(file_path)
        reader.Update()

        contour = vtkContourFilter()
        contour.SetInputConnection(reader.GetOutputPort())
        contour.SetComputeNormals(1)
        contour.SetComputeScalars(0)

        # Extract data range => Update store/state
        data_range = reader.GetOutput().GetPointData().GetScalars().GetRange()
        contour_value = 0.5 * (data_range[0] + data_range[1])
        state.data_range = data_range

        # Configure contour with valid values
        contour.SetNumberOfContours(1)
        contour.SetValue(0, contour_value)

        # Rendering setup
        renderer = vtkRenderer()
        renderWindow = vtkRenderWindow()
        renderWindow.AddRenderer(renderer)

        renderWindowInteractor = vtkRenderWindowInteractor()
        renderWindowInteractor.SetRenderWindow(renderWindow)
        renderWindowInteractor.GetInteractorStyle().SetCurrentStyleToTrackballCamera()

        mapper = vtkPolyDataMapper()
        actor = vtkActor()
        mapper.SetInputConnection(contour.GetOutputPort())
        actor.SetMapper(mapper)
        renderer.AddActor(actor)
        renderer.ResetCamera()
        # renderWindow.Render()

        self.renderWindow = renderWindow
        self.reader = reader







    def update_tab(self, tab, **kwargs):
        print(f"message===={tab}")
        self.message = tab


    def layout(self):
        # with RouterViewLayout(self.server, "/"):
        with vuetify.VCard():
            vuetify.VCardTitle("This is volume page")
            vuetify.VBtn(
                "volume view button",
                click=self.send_to_server1
            )

            vuetify.VBtn(
                "Volume On/Off",
                click=self.toggle_volume
            )
        with vuetify.VContainer(key='345'):
            view = vtk_widgets.VtkLocalView(self.renderWindow, v_show=('show_volume', False),
                                             ref="view_volume_1", style="min-height: 600px; max-width: 600px")
                # vuetify.VCardTitle("This is home")
            self.ctrl.view_update = view.update
            self.ctrl.view_reset_camera = view.reset_camera
            self.ctrl.on_server_ready.add(view.update)
       

    def toggle_volume(self):
        self.state.show_volume = not self.state.show_volume

```

**Reply by jourdain** [reply]
Since I only see pieces of the code, I also don't have any idea what you are doing wrong.
